kb(Form): keep form items order (#2095)
telerik/blazor-docs: knowledge-base/form-keep-fields-order.md

---
title: Keep Form Items Rendering Order after Show/Hide

> description: How to keep the form item rendering order upon toggling its visibility.

Reviewer: ```suggestion
description: How to keep the form item rendering order when changing the form item visibility.
```

type: how-to
page_title: Keep Form Items Rendering Order after Show/Hide
slug: form-kb-keep-fields-order
position: 
tags: 
ticketid: 1647539
res_type: kb
---

## Environment
<table>
	<tbody>
		<tr>
			<td>Product</td>
			<td>Form for Blazor</td>
		</tr>
	</tbody>
</table>

## Description

This knowledge base article answers the following questions:

* How do I keep the rendering order of Form items?
* Is it possible to toggle the visibility of a Form item and keep its rendering order instead of showing it as last?

## Solution

> The `<FormItems>` tag doesn't render Form items directly. This tag is used only to register `<FormItem>` instances. The Form finds out about the conditional field item last, so it puts it last.

Reviewer: I feel that some context might be lacking here. What is the "conditional field item" mentioned at the end (I'm guessing this is the form item that gets shown/hidden and moves around the form)? You probably need to provide some extra explanation, such as "form items whose visibility is subject to change are considered conditional field items" (or whatever makes sense).

> Thus, the solution is to use the `TelerikFormItemRenderer` as it defines the rendering place of each form item.

Reviewer: Would that be "instead of" or "along with"? 
```suggestion
To preserve the rendering place of each form item, you need to use the `TelerikFormItemRenderer` instead of the `<FormItems>` tag.
```

Author: Not technically correct so I adjusted it a little.

        @{
            var formItems = formContext.Items.Cast<IFormItem>();
        }
        <TelerikFormItemRenderer Item="@formItems.First(x => x.Field == nameof(FormModel.Field1))" />
        @if (model.Field1 == 2)
        {
            <TelerikFormItemRenderer Item="@formItems.First(x => x.Field == nameof(FormModel.Field2))" />
        }
        <TelerikFormItemRenderer Item="@formItems.First(x => x.Field == nameof(FormModel.Field3))" />
    </FormItemsTemplate>
</TelerikForm>

@code {
    private FormModel model { get; set; } = new FormModel();

    private List<RadioButtonItem> Items = new List<RadioButtonItem>()
    {
        new RadioButtonItem { ValueField = 1, TextField = "Item 1" },
        new RadioButtonItem { ValueField = 2, TextField = "Item 2" },
        new RadioButtonItem { ValueField = 3, TextField = "Item 3" },
    };

    public class FormModel
    {
        public int Field1 { get; set; }
        public int Field2 { get; set; }
        public int Field3 { get; set; }
    }

    public class RadioButtonItem
    {
        public string TextField { get; set; }
        public int ValueField { get; set; }
    }
}
````

## See Also

> * [Form Template for All Items]({%slug form-formitems-formitemstemplate%})

Reviewer: You might want to link to the other Form KBs that are available and to the Forms demo.
